### DISCUSSION TITLE
OpenVSCode and Laravel behind a reverse proxy

### DISCUSSION THREAD

**Discussion by Verhoeckx** [gitpod-io/openvscode-server#278] [Q&A]
Hello OpenVSCode developers!

Is there someone here who can help me to make the setup below work?

![Software setup - Sketch](https://user-images.githubusercontent.com/64807887/147661171-21e8b64b-265f-49c7-9ca0-4a2e33518421.png)

<br>

**Host**
Red Hat Enterprise Linux
Apache used as a reverse proxy

<br>

**Virtual machine**
Red Hat Enterprise Linux
Laravel is installed and reachable with HTTPS
Apache is used as the web server

<br>

Now I would like to run OpenVSCode is such away that it can be reached on port 3000 and that I can edit the content of the directory `/var/www/html` live:  I'm not sure this is possible, but I hope so :sweat_smile.

<br>

**Error**
When I now try to reach my.domain.com:3000/?tkn=[token] I get the message 'Unable to connect'.

<br>

**Note**
I opened port 3000 on both the host and the virtual machine with commands below.
```
sudo firewall-cmd --permanent --zone=public --add-port=3000/tcp
sudo firewall-cmd --reload
```

<br>

**Questions**
1. Does somebody know what the exact Apache configuration on the host must be?
2. Is there also a Apache configuration needed inside the virtual machine?
    As far as I know NodeJS has its own web server?
4. Is it allowed to mount the directory `/var/www/html` when running the container?

<br>

Al the steps I took so far, can be found here:
[https://jeroenverhoeckx.com/build-your-own-online-ide.html](https://jeroenverhoeckx.com/build-your-own-online-ide.html)

<br>

ps. I cleaned up the text of this issue to make it more readable.

**Comment by Verhoeckx** [comment]
Mmm, I realize that OpenVSCode needs read & write access to the directory `/var/www/html`.

Did this:
```
sudo semanage fcontext -a -t httpd_sys_rw_content_t '/var/www/html(/.*)?'
sudo restorecon -R '/var/www/html/'
```

**Comment by jeanp413** [comment]
@Verhoeckx for the reverse proxy configuration you can take a look at this discussion https://github.com/gitpod-io/openvscode-server/discussions/96#discussioncomment-1404354 hope that helps.

**Comment by Verhoeckx** [comment]
Should the lines below exist on the host or in the virtual host file on the virtual machine?

```
    RewriteEngine On
    RewriteCond %{HTTP:Connection} Upgrade [NC]
    RewriteCond %{HTTP:Upgrade} websocket [NC]
    RewriteRule /(.*) ws://192.168.122.237:3000/$1 [P,L]
```

**Comment by jeanp413** [comment]
I think it should be on the host, btw make sure you are using apache version 2.4.48 or greater as pointed in the linked discussion

**Reply by Verhoeckx** [reply]
Thanks for the quick reply!

I see that I am using version 2.4.37. Tomorrow I'm going to find how I can upgrade Apache.
```
$ httpd -v
Server version: Apache/2.4.37 (Red Hat Enterprise Linux)
Server built:   Oct 26 2021 14:18:06
```

**Comment by Verhoeckx** [comment]
I have good and bad news:

<br>

**Good news**
I added the lines below to the virtual host / reverse proxy:
```
RewriteEngine On
RewriteCond %{HTTP:Connection} Upgrade [NC]
RewriteCond %{HTTP:Upgrade} websocket [NC]
RewriteRule /(.*) ws://192.168.122.237:3000/$1 [P,L]
```

<br>

**Bad news**
I can't update Apache because 2.4.37 is the latest supported version by RHEL.

<br>

ps. I updated the image in this discussion.

**Comment by Verhoeckx** [comment]
I made some progress: I can now access OpenVSCode on port 3000!

<br>

![Screenshot 2021-12-30 at 13-14-23 Get Started - OpenVSCode Server](https://user-images.githubusercontent.com/64807887/147751142-534e8aaa-0797-40ea-b002-55b628873232.png)

<br>

**What I did**
I made Apache on the host, listen to port 3000 and then all traffic is proxied to the virtual machine.

<br>


Now I only get the error `The workbench failed to connect to the server (Error: WebSocket close with status code 1006)`.
So there is some kind of websocket problem.

I updated my online logboek:  
[https://jeroenverhoeckx.com/build-your-own-online-ide.html](https://jeroenverhoeckx.com/build-your-own-online-ide.html)

**Reply by jeanp413** [reply]
Awesome, that error is the same that others mentioned in the linked discussion, it's caused by the outdated apache server

**Reply by Verhoeckx** [reply]
So the good news is that my configuration is probably right, but the bad news is I can't update Apache because that's not supported.

Let's hope someone from Red Hat will answer my question about that (https://access.redhat.com/discussions/6616111).

@jeanp413: thanks for your help!!

**Comment by Verhoeckx** [comment]
One more question: do you know if it's allowed to mount the (public) directory `/var/www/html` to the container?

**Reply by jeanp413** [reply]
yeah, that should work, also ensure it has the right permissions

**Reply by Verhoeckx** [reply]
Thanks!!

Docker runs with root privileges so I think its allowed to read/write the files?
Or should, can I run docker as the user Apache (if that's possible)?

About the problem of the outdated Apache: I asked the same question on the mailing list of Apache. Hope that someone knows a workaround!

**Reply by jeanp413** [reply]
Running docker with root privileges should do it :+1:

**Comment by Verhoeckx** [comment]
I'm waiting from a second response from someone from Apache, but in the mean time I found out the following:

When I try to connect with:
```
sudo podman run -it --init -p 3000:3000 -v "$(pwd):/var/www/html:cached" gitpod/openvscode-server
```

<br>

And go to the address with the token
```
https:my.domain.com:3000/?tkn=[generated token]
```

<br>

I see the following error message in Firefox console tab:
```
Firefox can’t establish a connection to the server at wss://[my.domain.com]:3000/?reconnectionToken=[generated token]&reconnection=false&skipWebSocketFrames=false.
```

<br>

What is surprises me, is that it mentions `wss` while I don't have that in my Apache virtual host file / reverse proxy. How does it come with the `wss`?

**Reply by jeanp413** [reply]
Internally vscode web starts a `wss` or `ws` connection depending if you are using `https` or `http`

**Reply by Verhoeckx** [reply]
In my setup HTTPS is 'downgraded' (off-loaded) to HTTP in the proxy because SSL can't be configured in the virtual machine (it's now allowed and possible, as far as I know).

```
Internet  <-- HTTPS -->  Apache reverse proxy  <-- HTTP -->  Virtual machine
```

Somehow I need to tell OpenVSCode should use `ws` and not `wss`. Is that possible?

The HTTPS part is only to the host machine (this works with my Nextcloud vm).

https://[my.domain.com]:3000/?tkn=[security token]

<br>

My virtual host file is
```
<VirtualHost *:3000>
    ServerName [my.domain.com]

    RewriteEngine On
    RewriteCond %{HTTP:Upgrade} websocket [NC]
    RewriteCond %{HTTP:Connection} upgrade [NC]
    RewriteRule /(.*) ws://192.168.122.237:3000/$1 [P,L]

    ProxyPreserveHost On
    ProxyPass / http://192.168.122.237:3000/
    ProxyPassReverse / http://192.168.122.237:3000/

    SSLCertificateFile /etc/letsencrypt/live/[my.domain.com]/fullchain.pem
    SSLCertificateKeyFile /etc/letsencrypt/live/[my.domain.com]/privkey.pem
    Include /etc/letsencrypt/options-ssl-apache.conf
    SSLEngine on
</VirtualHost>
```

**Reply by jeanp413** [reply]
The config It's fine, apache will [forward the requests properly](https://stackoverflow.com/a/38852630)

**Comment by Verhoeckx** [comment]
I tried to use `--connection-secret [absolute path to the security token file]`, but when I did this, I got a `502 proxy error` every time.

I can't go back using this because I [already configured](https://jeroenverhoeckx.com/build-your-own-online-ide.html#start-on-boot) a systemd unit file based on the `--connection-token` argument.

**Comment by Verhoeckx** [comment]
I reported the problem at Red Hat's bugzilla!
